src/main/java/org/dsa/iot/alarm/inMemory/InMemoryProvider.java

```diff
@@ -0,0 +1,237 @@
+/* THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD
+ * TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN
+ * NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
+ * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
+ * IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+package org.dsa.iot.alarm.inMemory;
+
+import org.dsa.iot.alarm.*;
+import org.slf4j.*;
+import java.util.*;
+
+/**
+ * Alarming provider that stores alarms in memory using Java collections.
+ *
+ * @author Aaron Hansen
+ */
+public class InMemoryProvider extends AbstractProvider {
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Constants
+    ///////////////////////////////////////////////////////////////////////////
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(InMemoryProvider.class);
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Fields
+    ///////////////////////////////////////////////////////////////////////////
+
+    private static Comparator<AlarmRecord> alarmComparator = new AlarmComparator();
+    private Set<AlarmRecord> alarmCache = new TreeSet<>(alarmComparator);
+    private Map<UUID, AlarmRecord> alarmMap = new TreeMap<>();
+    private Set<AlarmRecord> alarmSet = new TreeSet<>(alarmComparator);
+    private List<Note> noteCache = new LinkedList<>();
+    private List<Note> noteList = new LinkedList<>();
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Constructors
+    ///////////////////////////////////////////////////////////////////////////
+
+    public InMemoryProvider() {
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Methods
+    ///////////////////////////////////////////////////////////////////////////
+
+    @Override public synchronized void addAlarm(AlarmRecord record) {
+        alarmSet.add(record);
+        alarmCache = new TreeSet<>(alarmComparator);
+        alarmMap.put(record.getUuid(),record);
+    }
+
+    @Override protected synchronized void addNote(Note note) {
+        noteList.add(note);
+        noteCache = new LinkedList<>();
+    }
+
+    @Override public void deleteAllRecords() {
+        alarmSet.clear();
+        alarmMap.clear();
+        noteList.clear();
+        alarmCache = new TreeSet<>(alarmComparator);
```

Comment:
There could be cursors working on the old cache, concurrent modification exceptions aren't worth it here.